### PR TITLE
Feature/geode 8768

### PR DESCRIPTION
**Pull request by mivanac**
Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [*] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [*] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [*] Is your initial contribution a single, squashed commit?

- [*] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

**Comment by kohlmu-pivotal** [comment]
@mivanac Is this a "work in progress" PR? If so, please change its status to "Draft".
Otherwise, @nabarunnag @agingade would is be possible to review these changes?

**Comment by mivanac** [comment]
This PR is ready to be reviewed.

**Review by mkevo — APPROVED** [review]

**Comment by pivotal-amurmann** [comment]
Adding folks who according to CODEOWNERS should review this.
@jinmeiliao @jdeppe-pivotal @kirklund 
@boglesby @gesterzhou 
@bschuchardt @Bill 

Wow, lots of owners for one PR!

**Review by bschuchardt — CHANGES_REQUESTED** [review]
The changes have left DistributionLocatorId in a bad state.  Its class javadoc needs to be updated to include hostNameForClients and memberName, and the equality-check methods need to be documented.  The name "additionalEqualityCheck", in particular, seems meaningless to me and I had to read its code and look at its uses in order to figure out what it does.  A javadoc will help but I think a more meaningful name is needed for this new method.

File: geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/misc/WanLocatorDiscoveryDUnitTest.java
Comment (jinmeiliao) [inline geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/misc/WanLocatorDiscoveryDUnitTest.java:70]: You don't need all these when you use ClusterStartupRule. The rule sets those properties for you already so that you can avoid all these boiler code.

The rule also allocates the available ports for you so that you don't need to pre-allocate them.

use something like:

locator = cluster.startLocatorVM(index, l->l.withPropety(....).withConnectionToLocator(port0, port1, port2));

**Review by jinmeiliao — CHANGES_REQUESTED** [review]

**Review by bschuchardt — APPROVED** [review]
Thank you Mario

File: geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/misc/WanLocatorDiscoveryDUnitTest.java
Comment (jinmeiliao) [inline geode-wan/src/distributedTest/java/org/apache/geode/internal/cache/wan/misc/WanLocatorDiscoveryDUnitTest.java:69]: you can get rid of `withProperty(MCAST_PORT, "0")`, it's already set by the rule

**Review by jinmeiliao — APPROVED** [review]

**Review by bschuchardt — APPROVED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/admin/remote/DistributionLocatorId.java
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/internal/admin/remote/DistributionLocatorId.java:137]: is this "::" a typo here?

File: geode-core/src/main/java/org/apache/geode/internal/admin/remote/DistributionLocatorId.java
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/internal/admin/remote/DistributionLocatorId.java:131]: it's better to put an example of membername here.

File: geode-core/src/main/java/org/apache/geode/internal/admin/remote/DistributionLocatorId.java
Comment (gesterzhou) [inline geode-core/src/main/java/org/apache/geode/internal/admin/remote/DistributionLocatorId.java:375]: I think it should be "||", not "&&".

**Review by gesterzhou — CHANGES_REQUESTED** [review]

File: geode-core/src/main/java/org/apache/geode/internal/admin/remote/DistributionLocatorId.java
Comment (gesterzhou) [reply geode-core/src/main/java/org/apache/geode/internal/admin/remote/DistributionLocatorId.java:375]: Bruce suggested to use:

if (!Objects.equals(this.host, that.host) {
  return false;
}

**Review by gesterzhou — APPROVED** [review]

**Comment by ladyVader** [comment]
It seems that DistributionLocatorId.equals() can return true if one of the two compared objects doesn't have the memberName and the other does ... but DistributionLocatorId.hashCode() will be different ... if we invoke contains() on a Set of DistributionLocatorIds ... contains() will return false in that case.

I also found the logging a bit misleading once toString() didn't include the memberName.

**Comment by mivanac** [comment]
Equals() and hashCode(), are modified to support cases, when sites with old implementation are discovered with new locators. Also this is the reason why membername is not included in toString().

**Comment by mivanac** [comment]
If this is creating problems in your tests, I will need to modify logic.

**Comment by bschuchardt** [comment]
@mivanac I think the best thing to do at this point is revert the merge and write some tests to fix the problems with

1. membername being an empty string.  In this case all other fields are ignored.
2. hashcode returning a value that places IDs having/not-having membernames in different hashmap buckets

I noticed #1 when reviewing code when people were talking about #2.  I wrote a test for #1.  A test for #2 is harder to write.

DistributionLocatorIdJUnitTest.java
```java
@Test
public void testEqualsWithMemberId() throws UnknownHostException {
  InetAddress address = InetAddress.getLocalHost();
  DistributionLocatorId dLI1 = new DistributionLocatorId(address, 40404, "127.0.0.1", null);
  DistributionLocatorId dLI2 = dLI1;
  @SuppressWarnings("RedundantStringConstructorCall")
  DistributionLocatorId dLI3 =
      new DistributionLocatorId(address, 40404, new String("127.0.0.1"), null);
  @SuppressWarnings("RedundantStringConstructorCall")
  DistributionLocatorId dLI4 = new DistributionLocatorId(InetAddress.getByName("localhost"),
      40405, new String("127.0.0.1"), null);
  dLI1.setMembername("");
  dLI3.setMembername("");
  dLI4.setMembername("");
  assertTrue(dLI1.equals(dLI2));
  assertTrue(dLI1.equals(dLI3));
  assertFalse(dLI1.equals(dLI4));
}
```

**Pull request merged**
